### ISSUE TITLE
Tables unable to repair

### ISSUE THREAD

**Issue by zooptwopointone** [scylladb/scylladb#8022]
I have also run into something like this. Cluster info:
Scylla version 4.2.1-0.20201108.4fb8ebccff with build-id d19a95fd85a8e7df2928ccd8c729941f263e8adc
Cluster size: 4 Datacenters (3 x 3 nodes 1 x 7 nodes)
OS (RHEL/CentOS/Ubuntu/AWS AMI): 18.04.5 LTS (GNU/Linux 4.15.0-128-generic x86_64)

Platform (physical/VM/cloud instance type/docker): Physical
Hardware: sockets= 2 cores=8 hyperthreading= 32 memory= 128 GB
Disks: (SSD/HDD, count) 8 SSD

I have been trying to run repairs on tables, but their are a few that just will not complete successfully.  While it is not failing on every node on the ones it does fail on it seems so far consistant for ones that will not repair.

Sometimes I get these same failures and I just run a major compaction on the table then run the repair again and a lot of the time it will then work.  but sometimes not.  Here I have pasted the output of the nodetool result as well as the log final result from Shard 0

Let me know what other information you might need.

# nodetool repair -pr  -- keyspace table_status
[2021-01-29 12:26:24,667] Starting repair command #4, repairing 1 ranges for keyspace keyspace (parallelism=SEQUENTIAL, full=true)
[2021-01-29 12:35:50,848] Repair session 4 failed
[2021-01-29 12:35:50,849] Repair session 4 finished
error: Repair job has failed with the error message: [2021-01-29 12:35:50,848] Repair session 4 failed
-- StackTrace --
java.lang.RuntimeException: Repair job has failed with the error message: [2021-01-29 12:35:50,848] Repair session 4 failed
        at org.apache.cassandra.tools.RepairRunner.progress(RepairRunner.java:124)
        at org.apache.cassandra.utils.progress.jmx.JMXNotificationProgressListener.handleNotification(JMXNotificationProgressListener.java:77)
        at com.sun.jmx.remote.internal.ClientNotifForwarder$NotifFetcher.dispatchNotification(ClientNotifForwarder.java:583)
        at com.sun.jmx.remote.internal.ClientNotifForwarder$NotifFetcher.doRun(ClientNotifForwarder.java:533)
        at com.sun.jmx.remote.internal.ClientNotifForwarder$NotifFetcher.run(ClientNotifForwarder.java:452)
        at com.sun.jmx.remote.internal.ClientNotifForwarder$LinearExecutor$1.run(ClientNotifForwarder.java:108)

Jan 29 12:35:50 id2296gc scylla[26762]:  [shard 0] repair - repair id 4 failed: std::runtime_error ({shard 7: std::runtime_error (Failed to generate sharding metadata for /var/lib/scylla/data/keyspace/table_status-821bbf40e79511e89e8c000000000001/mc-338239-big-Data.db)})
Jan 29 12:35:50 id2296gc scylla[26762]:  [shard 0] repair - repair_tracker run for repair id 4 failed: std::runtime_error ({shard 7: std::runtime_error (Failed to generate sharding metadata for /var/lib/scylla/data/keyspace/table_status-821bbf40e79511e89e8c000000000001/mc-338239-big-Data.db)})

**Comment by slivne** [comment]
@zooptwopointone 
- lets get the info on your cluster setup on the other issue #8023 

Please note we did find a bug related to aborted repairs https://github.com/scylladb/scylla/issues/7411 that may cause issues and the fix for this was released in 4.2.3. I am not sure its related to what you are seeing - but I do recommend that your upgrade especially if you are having issues in repair - to make sureyou do not hit this.

@denesb / @bhalevy - the only case I found `Failed to generate sharding metadata for` was in unitest failure in 3.2 and it hasn't reproduced since then - any thoughts ?

**Comment by denesb** [comment]
The error `Failed to generated sharding metadata` is raised when the sstable is empty and hence there is no partition range(s) in it to calculate the sharding metadata based on. On a related note we should add a dedicated check to be able to raise a more clear error message.

**Comment by slivne** [comment]
@bhalevy I tried to search I think we had a case in the past in which we created an sstable in some cases without it having data - I also checked with @zooptwopointone they have a cluster with nodes with a different shard count and running repairs - could it be somehow related.

I know we have discussed this in the past - and sstables are supposed to be open lazily - when there is a first write - but maybe we are missing a case - lets recheck this - we are seeing more people complain about this.

**Comment by bhalevy** [comment]
@asias can repair (when the repair master has different number of shards than the repair follower) write empty sstables?

**Comment by asias** [comment]
The writer on each shard is created when there is at least one mutation fragment belongs to this shard.  @denesb will such error be seen if we only have fragment start and fragment end, no other fragments in the middle?

```

170 future<> multishard_writer::distribute_mutation_fragments() {
171     return repeat([this] () mutable {
172         return _producer(db::no_timeout).then([this] (mutation_fragment_opt mf_opt) mutable {
173             if (mf_opt) {
174                 return handle_mutation_fragment(std::move(*mf_opt));
175             } else {
176                 return handle_end_of_stream();
177             }
178         });
179     }).handle_exception([this] (std::exception_ptr ep) {
180         for (auto& q : _queue_reader_handles) {
181             if (q) {
182                 q->abort(ep);
183             }
184         }
185         return make_exception_future<>(std::move(ep));
186     });
187 }

108 future<> multishard_writer::make_shard_writer(unsigned shard) {
109     auto [reader, handle] = make_queue_reader(_s, _producer.permit());
110     _queue_reader_handles[shard] = std::move(handle);
111     return smp::submit_to(shard, [gs = global_schema_ptr(_s),
112             consumer = _consumer,
113             reader = make_foreign(std::make_unique<flat_mutation_reader>(std::move(reader)))] () mutable {
114         auto s = gs.get();
115         auto this_shard_reader = make_foreign_reader(s, semaphore.make_permit(s.get(), "multishard-writer"), std::move(reader));
116         return make_foreign(std::make_unique<shard_writer>(gs.get(), std::move(this_shard_reader), consumer));
117     }).then([this, shard] (foreign_ptr<std::unique_ptr<shard_writer>> writer) {
118         _shard_writers[shard] = std::move(writer);
119         _pending_consumers.push_back(consume(shard));
120     });
121 }
122 
123 future<stop_iteration> multishard_writer::handle_mutation_fragment(mutation_fragment mf) {
124     auto f = make_ready_future<>();
125     if (mf.is_partition_start()) {
126         _consumed_partitions++;
127         if (unsigned shard = shard_for_mf(mf); shard != _current_shard) {
128             _current_shard = shard;
129             if (!bool(_shard_writers[shard])) {
130                 f = make_shard_writer(shard);
131             }
132         }
133     }
134     return f.then([this, mf = std::move(mf)] () mutable {
135         assert(_current_shard != -1u);
136         return _queue_reader_handles[_current_shard]->push(std::move(mf));
137     }).then([] {
138         return stop_iteration::no;
139     });
140 }
```

**Comment by denesb** [comment]
I think even an empty partition should be enough to make this go away, but I'm not 100% sure, @raphaelsc can you advise?

**Comment by avikivity** [comment]
Reopening, looks like the wrong issue was references in the commit.

@denesb what is the correct issue?

**Comment by slivne** [comment]
@avikivity I already fixed that ....

**Comment by slivne** [comment]
It is https://github.com/scylladb/scylla/issues/8035

**Comment by denesb** [comment]
Looks like I copied the wrong issue number, sorry for the noise.

**Comment by raphaelsc** [comment]
> The error `Failed to generated sharding metadata` is raised when the sstable is empty and hence there is no partition range(s) in it to calculate the sharding metadata based on. On a related note we should add a dedicated check to be able to raise a more clear error message.

Before generating sharding metadata, we retrieve both first and last decorated key, and this will only succeed if they aren't empty. 
So "Failed to generate sharding metadata" means that the key range spanned by the sstable belongs to another shard.

See this patch of mine which introduced this exception: https://github.com/scylladb/scylla/commit/fa5a26f12d69c1e1a9a9689308f84ccd6aa54f34

So we need to investigate how data was feeded to the wrong shard. IIRC, in the past, that happened when view table used incorrect schema (https://github.com/scylladb/scylla/commit/115ff1095edc59097fc599e491be8a3538366bb7), and its impact can be seen here: https://github.com/scylladb/scylla/issues/2932

**Comment by raphaelsc** [comment]
@zooptwopointone please share schema of keyspace/table_status

**Comment by zooptwopointone** [comment]
CREATE TABLE keyspace.table_status (
    sid uuid,
    contact_id uuid,
    campaign_id uuid,
    campaign_run_id uuid,
    campaign_path list<int>,
    data map<text, text>,
    failure_reason text,
    last_phone text,
    next_node int,
    status set<int>,
    workflow_version int,
    PRIMARY KEY ((sid, contact_id), campaign_id, campaign_run_id)
) WITH CLUSTERING ORDER BY (campaign_id ASC, campaign_run_id ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'ALL'}
    AND comment = ''
    AND compaction = {'class': 'SizeTieredCompactionStrategy'}
    AND compression = {'sstable_compression': 'org.apache.cassandra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99.0PERCENTILE';

**Comment by slivne** [comment]
@raphaelsc ping

**Comment by raphaelsc** [comment]
so there's no segregation enabled given that STCS is used.

the problem is likely in multishard_writer or a bad interaction between it and row level repair, as a new sstable in shard S was feeded data that didn't belong to it, sstable wasn't empty as originally thought because other exception would've been triggered instead.

**Comment by zooptwopointone** [comment]
Please let me know if there is something I an provide to help this along.

**Comment by zooptwopointone** [comment]
I ran the repair running 4.2.3 and got these errors.

relavant syslog data : 

```
Jun 04 17:49:30 ID1888LA scylla[4413]:  [shard 0] repair - repair id [id=1, uuid=c6ea4ef9-ad8b-4413-9f1d-5d71b91ab092] failed: std::runtime_error ({shard 5: std::runtime_error (Failed to generate sharding metadata for /var/lib/scylla/data/keyspace/table_status-821bbf40e79511e89e8c000000000001/md-1140425-big-Data.db)})
Jun 04 17:49:30 ID1888LA scylla[4413]:  [shard 0] repair - repair_tracker run for repair id [id=1, uuid=c6ea4ef9-ad8b-4413-9f1d-5d71b91ab092] failed: std::runtime_error ({shard 5: std::runtime_error (Failed to generate sharding metadata for /var/lib/scylla/data/keyspace/table_status-821bbf40e79511e89e8c000000000001/md-1140425-big-Data.db)})
```

command:

```
# nodetool repair -pr -local -- keyspace table_status
Using /etc/scylla/scylla.yaml as the config file
[2021-06-04 17:42:22,076] Starting repair command #1, repairing 1 ranges for keyspace keyspace (parallelism=SEQUENTIAL, full=true)
[2021-06-04 17:49:31,278] Repair session 1 failed
[2021-06-04 17:49:31,278] Repair session 1 finished
error: Repair job has failed with the error message: [2021-06-04 17:49:31,278] Repair session 1 failed
-- StackTrace --
java.lang.RuntimeException: Repair job has failed with the error message: [2021-06-04 17:49:31,278] Repair session 1 failed
	at org.apache.cassandra.tools.RepairRunner.progress(RepairRunner.java:124)
	at org.apache.cassandra.utils.progress.jmx.JMXNotificationProgressListener.handleNotification(JMXNotificationProgressListener.java:77)
	at com.sun.jmx.remote.internal.ClientNotifForwarder$NotifFetcher.dispatchNotification(ClientNotifForwarder.java:583)
	at com.sun.jmx.remote.internal.ClientNotifForwarder$NotifFetcher.doRun(ClientNotifForwarder.java:533)
	at com.sun.jmx.remote.internal.ClientNotifForwarder$NotifFetcher.run(ClientNotifForwarder.java:452)
	at com.sun.jmx.remote.internal.ClientNotifForwarder$LinearExecutor$1.run(ClientNotifForwarder.java:108)

```

**Comment by raphaelsc** [comment]
After analyzing the data and reading relevant code, I think you're potentially affected by https://github.com/scylladb/scylla/commit/289a08072ac59b49504141ebf0b7b3b000f28f66, which is fixed by https://github.com/scylladb/scylla/commit/85bbf6751d10006d9b05fdd77955cd7e9911e9d3, and only available in scylla 4.3.3. Fix was also backported to branch-4.2 but it's still not available in any 4.2 minor release.

**Comment by raphaelsc** [comment]
> After analyzing the data and reading relevant code, I think you're potentially affected by [289a080](https://github.com/scylladb/scylla/commit/289a08072ac59b49504141ebf0b7b3b000f28f66), which is fixed by [85bbf67](https://github.com/scylladb/scylla/commit/85bbf6751d10006d9b05fdd77955cd7e9911e9d3), and only available in scylla 4.3.3. Fix was also backported to branch-4.2 but it's still not available in any 4.2 minor release.

/CC @zooptwopointone

**Comment by asias** [comment]
fixed in https://github.com/scylladb/scylladb/commit/85bbf6751d10006d9b05fdd77955cd7e9911e9d3
